### PR TITLE
[FLINK-12098] [table-planner-blink] Add support for generating optimized logical plan for simple group aggregate on stream

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*Add support for generating optimized logical plan for simple group aggregate on stream*

## Brief change log

  - *add StreamExecGroupAggregateRule to convert logical aggregate to StreamExecGroupAggregate*
  - *add TwoStageOptimizedAggregateRule to write StreamExecGroupAggregate to two-stage aggregates*
  - *add StreamExecRetractionRules to handle retraction message*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added AggregateTest for StreamExecGroupAggregateRule*
  - *Added TwoStageAggregateTest for TwoStageOptimizedAggregateRule*
  - *Added RetractionRulesTest for retraction rules*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/runtime/functions/aggfunctions/MaxAggFunctionWithRetract.scala [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+import org.apache.flink.table.functions.AggregateFunction
+import org.apache.flink.table.runtime.functions.aggfunctions.Ordering._
+import org.apache.flink.table.typeutils.DecimalTypeInfo
+
+import java.lang
+import java.lang.{Iterable => JIterable, Long => JLong}
+import java.sql.{Date, Time, Timestamp}
+
+/** The initial accumulator for Max with retraction aggregate function */
+class MaxWithRetractAccumulator[T] {
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/runtime/functions/aggfunctions/MaxAggFunctionWithRetract.scala]: Can we make this Java?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/runtime/functions/aggfunctions/MinAggFunctionWithRetract.scala [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+import org.apache.flink.table.functions.AggregateFunction
+import org.apache.flink.table.runtime.functions.aggfunctions.Ordering._
+import org.apache.flink.table.typeutils.DecimalTypeInfo
+
+import java.lang
+import java.lang.{Iterable => JIterable, Long => JLong}
+import java.sql.{Date, Time, Timestamp}
+
+/** The initial accumulator for Min with retraction aggregate function */
+class MinWithRetractAccumulator[T] {
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/runtime/functions/aggfunctions/MinAggFunctionWithRetract.scala]: Can we make this Java?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/AggregateUtil.scala [hunk 3]
```diff
@@ -495,4 +497,65 @@ object AggregateUtil extends Enumeration {
...
+
+  /**
+    * Optimize max or min with retraction agg. MaxWithRetract can be optimized to Max if input is
+    * update increasing.
+    */
+  def getNeedRetractions(
+      groupSize: Int,
+      needRetraction: Boolean,
+      aggs: Seq[AggregateCall]): Array[Boolean] = {
+
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/AggregateUtil.scala]: delete blank line

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/AggregateUtil.scala [hunk 4]
```diff
@@ -495,4 +497,65 @@ object AggregateUtil extends Enumeration {
...
+  /**
+    * Derives output row type from local aggregate
+    */
+  def inferLocalAggRowType(
+      aggInfoList: AggregateInfoList,
+      inputRowType: RelDataType,
+      groupSet: Array[Int],
+      typeFactory: FlinkTypeFactory): RelDataType = {
+
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/AggregateUtil.scala]: delete blank line

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/AggregateUtil.scala [hunk 5]
```diff
@@ -495,4 +497,65 @@ object AggregateUtil extends Enumeration {
...
+    typeFactory.buildRelDataType(
+      groupingNames ++ accFieldNames,
+      groupingTypes ++ accTypes.map(TypeConverters.createInternalTypeFromTypeInfo))
+  }
+
+  /**
+    * Derives accumulators names from aggregate
+    */
+  def inferAggAccumulatorNames(aggInfoList: AggregateInfoList): Array[String] = {
+
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/AggregateUtil.scala]: delete blank line

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/runtime/functions/aggfunctions/MaxAggFunctionWithRetract.scala (hunk 1 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/runtime/functions/aggfunctions/MaxAggFunctionWithRetract.scala]: ok, should implements MinWithRetractAggFunctionTest as java in this PR?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/functions/aggfunctions/MaxWithRetractAggFunction.java [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+/**
+ * built-in Max with retraction aggregate function.
+ */
+public abstract class MaxWithRetractAggFunction<T>
+		extends AggregateFunction<T, MaxWithRetractAggFunction.MaxWithRetractAccumulator<T>> {
+
+	/** The initial accumulator for Max with retraction aggregate function. */
+	public static class MaxWithRetractAccumulator<T> {
+		public T max;
+		public Long distinctCount;
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/functions/aggfunctions/MaxWithRetractAggFunction.java]: OK

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/functions/aggfunctions/MaxWithRetractAggFunction.java [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+		public MapView<T, Long> map;
+	}
+
+	@Override
+	public MaxWithRetractAccumulator<T> createAccumulator() {
+		MaxWithRetractAccumulator<T> acc = new MaxWithRetractAccumulator<>();
+		acc.max = getInitValue(); // max
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/functions/aggfunctions/MaxWithRetractAggFunction.java]: I does not change any logic when porting to java.
for java, using NULL as init value is ok.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/functions/aggfunctions/MaxWithRetractAggFunction.java [hunk 8]
```diff
@@ -0,0 +1,482 @@
...
+					boolean hasMax = false;
+					Comparator<T> comparator = getComparator();
+					while (iterator.hasNext()) {
+						T key = iterator.next();
+						if (!hasMax || comparator.compare(acc.max, key) < 0) {
+							acc.max = key;
+							hasMax = true;
+						}
+					}
+					if (!hasMax) {
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/functions/aggfunctions/MaxWithRetractAggFunction.java]: The behavior of deleting expired data in the state backend is uncertain. 
so `mapSize` data may exist while `map` data may have been deleted when both of them are expired.

**Comment by KurtYoung** [comment]
LGTM, +1 to merge

**Pull request merged**